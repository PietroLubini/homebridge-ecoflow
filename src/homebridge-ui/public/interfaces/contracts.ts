import { IHomebridge, PluginConfig, PluginConfigSchema } from '../interfaces/homebridge';

export interface IContext {
  homebridgeProvider: IHomebridge;
  configSchema: PluginConfigSchema;
  configuration: PluginConfig;
}

export interface IDeviceContext extends IContext {
  hideDeviceSettingsPerModel: Record<string, string[]>;
}

export type ChangedCallbackHandler = (value: string) => void;

export interface IRenderer {
  render(context: IContext): void;
}
