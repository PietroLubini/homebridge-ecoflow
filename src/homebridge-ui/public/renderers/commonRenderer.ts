import { ChangedCallbackHandler } from '../interfaces/contracts';
import { PluginConfigSchemaEnum, PluginConfigSchemaObject } from '../interfaces/homebridge';

export class CommonRenderer {
  public renderTextBox(
    $parent: JQuery,
    id: string,
    schemaProperty: PluginConfigSchemaObject,
    value: string,
    onChangeCallback: ChangedCallbackHandler
  ): void {
    const template = `
    <div class="form-group">
      <label for="device{id}" class="{requiredClass}">{title}</label>
      <input type="text" class="form-control" id="device{id}" placeholder="{description}" required="{required}" />
    </div>`;

    const html = template
      .replace(/{id}/g, id)
      .replace(/{title}/g, schemaProperty.title)
      .replace(/{description}/g, schemaProperty.description ?? '')
      .replace(/{requiredClass}/g, schemaProperty.required ? 'required-label' : '')
      .replace(/{required}/g, (schemaProperty.required ?? false).toString());
    $parent.append(html);

    const $control = $(`#device${id}`, $parent);
    $control.val(value);
    $control.on('change', () => onChangeCallback($control.val()?.toString() || ''));
  }

  public renderDropDown(
    $parent: JQuery,
    id: string,
    schemaProperty: PluginConfigSchemaEnum,
    value: string,
    onChangeCallback: ChangedCallbackHandler
  ): void {
    const template = `
    <div class="form-group">
      <label for="device{id}" class="{requiredClass}">{title}</label>
      <select class="form-control" id="device{id}" required="{required}"></select>
    </div>`;
    const html = template
      .replace(/{id}/g, id)
      .replace(/{title}/g, schemaProperty.title)
      .replace(/{requiredClass}/g, schemaProperty.required ? 'required-label' : '')
      .replace(/{required}/g, (schemaProperty.required ?? false).toString());
    $parent.append(html);

    const $control = $(`#device${id}`, $parent);
    schemaProperty.enum.forEach((enumValue: string) => {
      const $option = $('<option>').val(enumValue).text(enumValue);
      $control.append($option);
    });
    if (value !== undefined) {
      $control.val(value);
    }
    $control.on('change', () => onChangeCallback($control.val()?.toString() || ''));
  }
}
