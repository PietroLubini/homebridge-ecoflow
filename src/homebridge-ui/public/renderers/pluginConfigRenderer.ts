import { IContext, IRenderer } from '../interfaces/contracts';

export class PluginConfigRenderer {
  constructor(private readonly renderers: IRenderer[]) {}

  public render(context: IContext): void {
    if (context.configuration === undefined) {
      context.configuration = {
        name: undefined as unknown as string,
        devices: [],
      };
    }
    this.renderers.forEach(renderer => renderer.render(context));
  }
}
