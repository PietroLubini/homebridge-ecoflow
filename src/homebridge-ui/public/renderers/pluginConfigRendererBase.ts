import { IContext, IRenderer } from '../interfaces/contracts';
import { ComponentRenderer } from './componentRenderer';

export abstract class PluginConfigRendererBase implements IRenderer {
  constructor(protected readonly componentRenderer: ComponentRenderer) {}

  public abstract render(context: IContext): void;

  protected async updatePluginConfig(context: IContext): Promise<void> {
    const changes = this.clone(context.configuration);
    delete changes.platform;
    await context.homebridgeProvider.updatePluginConfig([changes]);
  }

  protected clone<TObject>(obj: TObject): TObject {
    return JSON.parse(JSON.stringify(obj));
  }
}
