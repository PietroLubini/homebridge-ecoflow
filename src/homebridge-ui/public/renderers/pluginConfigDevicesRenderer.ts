import { IContext, IDeviceContext } from '../interfaces/contracts';
import {
  IForm,
  PluginConfigSchema,
  PluginConfigSchemaDevicesItems,
  PluginConfigSchemaEnum,
  PluginDeviceConfig,
} from '../interfaces/homebridge';
import { ComponentRenderer } from './componentRenderer';
import { PluginConfigRendererBase } from './pluginConfigRendererBase';

export class PluginConfigDevicesRenderer extends PluginConfigRendererBase {
  private form: IForm | undefined;
  private readonly $tabs: JQuery;
  private readonly $tabPanels: JQuery;
  private readonly _hideDeviceSettingsPerModel: Record<string, string[]>;

  constructor(componentRenderer: ComponentRenderer) {
    super(componentRenderer);
    this.$tabs = $('#devicesTabs');
    this.$tabPanels = $('#devicesPanels');
    this._hideDeviceSettingsPerModel = {
      'Delta 2': ['powerStream', 'powerOcean', 'outlet'],
      'Delta 2 Max': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro 3': ['powerStream', 'powerOcean', 'outlet'],
      'Delta Pro Ultra': ['powerStream', 'powerOcean', 'outlet'],
      PowerStream: ['battery', 'powerOcean', 'outlet'],
      PowerOcean: ['battery', 'powerStream', 'outlet'],
      'Smart Plug': ['battery', 'powerStream', 'powerOcean'],
      Glacier: ['powerStream', 'powerOcean', 'outlet'],
    };
  }

  public get hideDeviceSettingsPerModel(): Record<string, string[]> {
    return this._hideDeviceSettingsPerModel;
  }

  public render(context: IContext): void {
    this.renderDevicesSettings({
      homebridgeProvider: context.homebridgeProvider,
      configSchema: context.configSchema,
      configuration: context.configuration,
      hideDeviceSettingsPerModel: this.hideDeviceSettingsPerModel,
    });
  }

  private renderDevicesSettings(context: IDeviceContext, activeDeviceIndex?: number): void {
    $('#devicesLabel').text(context.configSchema.schema.properties['devices'].title);
    const activeIndex = !activeDeviceIndex || activeDeviceIndex < 0 ? 0 : activeDeviceIndex;
    this.renderDeviceTabs(context, activeIndex);
  }

  private renderDeviceTabs(context: IDeviceContext, activeIndex: number): void {
    this.clearDeviceTabs();

    const tabTemplate = `
      <li class="nav-item">
        <a class="nav-link {activeClass}" data-toggle="tab" href="#deviceTabPanel{index}" id="deviceTab{index}">{name}</a>
      </li>
    `;

    context.configuration.devices.forEach((deviceConfiguration: PluginDeviceConfig, index: number) => {
      const activeTabClass = index === activeIndex ? 'active' : '';

      const tabHtml = tabTemplate
        .replace(/{activeClass}/g, activeTabClass)
        .replace(/{index}/g, index.toString())
        .replace(/{name}/g, deviceConfiguration.name ?? `Device${index + 1}`);

      this.$tabs.append(tabHtml);

      this.renderDeviceTabPanel(
        context,
        deviceConfiguration,
        index,
        activeIndex,
        this.$tabs.find(`#deviceTab${index}`)
      );
    });

    this.renderAddNewDeviceTab(tabTemplate, context);
  }

  private clearDeviceTabs(): void {
    this.$tabs.empty();
    this.$tabPanels.empty();
    this.form?.end();
  }

  private renderAddNewDeviceTab(template: string, context: IDeviceContext): void {
    const html = template
      .replace(/{activeClass}/g, '')
      .replace(/{index}/g, 'Add')
      .replace(/{name}/g, 'Add Device');
    this.$tabs.append(html);
    const $addButton = this.$tabs.find('#deviceTabAdd');
    $addButton.on('click', () => {
      context.configuration.devices.push(this.getDefaultDeviceConfiguration(context.configSchema));
      this.renderDevicesSettings(context, context.configuration.devices.length - 1);
    });
  }

  private renderDeviceTabPanel(
    context: IDeviceContext,
    deviceConfiguration: PluginDeviceConfig,
    index: number,
    activeIndex: number,
    $tab: JQuery
  ): void {
    const tabPanelTemplate = `
    <div class="tab-pane container fade {activeClass} card card-body list-group-item" id="deviceTabPanel{index}">
      <button type="button" class="close pull-right" id="deviceTabPanelClose{index}">
        <span>Ã—</span>
        <span class="sr-only">Close</span>
      </button>
    </div>
  `;

    const activeTabPanelClass = index === activeIndex ? 'in show active' : '';
    const tabPanelHtml = tabPanelTemplate
      .replace(/{activeClass}/g, activeTabPanelClass)
      .replace(/{index}/g, index.toString());

    this.$tabPanels.append(tabPanelHtml);
    const $tabPanel = this.$tabPanels.find(`#deviceTabPanel${index}`);

    const deviceSchema = this.clone(context.configSchema.schema.properties.devices.items);
    const modelSchema = deviceSchema.properties.model;
    this.renderModelDropDown($tab, $tabPanel, index, context, deviceSchema, modelSchema!, deviceConfiguration);
    this.renderRemoveButton($tabPanel, index, context);
    this.renderForm($tab, index, activeIndex, context, deviceSchema, deviceConfiguration);
  }

  private renderModelDropDown(
    $tab: JQuery,
    $tabPanel: JQuery,
    index: number,
    context: IDeviceContext,
    deviceSchema: PluginConfigSchemaDevicesItems,
    modelSchema: PluginConfigSchemaEnum,
    deviceConfiguration: PluginDeviceConfig
  ) {
    this.componentRenderer.renderDropDown(
      $tabPanel,
      'model' + index,
      modelSchema,
      deviceConfiguration.model,
      newValue => {
        deviceConfiguration.model = newValue;
        this.renderForm($tab, index, index, context, deviceSchema, deviceConfiguration);
      }
    );
  }

  private renderRemoveButton($parent: JQuery, index: number, context: IDeviceContext) {
    const $removeButton = $parent.find(`#deviceTabPanelClose${index}`);
    $removeButton.on('click', () => {
      context.configuration.devices.splice(index, 1);
      this.renderDevicesSettings(context, index - 1);
    });
  }

  private renderForm(
    $tab: JQuery,
    index: number,
    activeIndex: number,
    context: IDeviceContext,
    deviceSchema: PluginConfigSchemaDevicesItems,
    deviceConfiguration: PluginDeviceConfig
  ): void {
    deviceSchema = this.clone(deviceSchema);
    delete deviceSchema.properties.model;
    context.hideDeviceSettingsPerModel[deviceConfiguration.model].forEach(propertyName => {
      delete deviceSchema.properties[propertyName];
    });

    $tab.off('click').on('click', () => {
      this.form?.end();
      this.form = context.homebridgeProvider.createForm({ schema: deviceSchema }, deviceConfiguration);
      this.form.onChange(async newDeviceConfiguration => {
        const oldName = context.configuration.devices[index].name;
        this.applyChanges(context.configuration.devices[index], newDeviceConfiguration, ['model']);
        await this.updatePluginConfig(context);

        if (newDeviceConfiguration.name !== oldName) {
          $tab.text(newDeviceConfiguration.name);
        }
      });
    });
    if (index === activeIndex) {
      $tab.trigger('click');
    }
  }

  private getDefaultDeviceConfiguration(configSchema: PluginConfigSchema): PluginDeviceConfig {
    const result = {} as PluginDeviceConfig;
    const properties = configSchema.schema.properties.devices.items.properties;
    Object.keys(properties)
      .filter(propertyName => !!properties[propertyName]?.default)
      .forEach(propertyName => {
        result[propertyName] = properties[propertyName]?.default;
      });
    return result;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private applyChanges(target: any, source: any, readOnlyProperties: string[]): void {
    Object.assign(target, source);
    for (const key in target) {
      if (source[key] === undefined && !readOnlyProperties.includes(key)) {
        delete target[key];
      }
    }
  }
}
