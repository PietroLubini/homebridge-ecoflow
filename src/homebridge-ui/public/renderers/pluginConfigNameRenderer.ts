import { IContext } from '../interfaces/contracts';
import { PluginConfigRendererBase } from './pluginConfigRendererBase';

export class PluginConfigNameRenderer extends PluginConfigRendererBase {
  public render(context: IContext): void {
    this.componentRenderer.renderTextBox(
      $('#generalSettings'),
      'name',
      context.configSchema.schema.properties.name,
      context.configuration.name,
      async newValue => {
        context.configuration.name = newValue;
        await this.updatePluginConfig(context);
      }
    );
  }
}
