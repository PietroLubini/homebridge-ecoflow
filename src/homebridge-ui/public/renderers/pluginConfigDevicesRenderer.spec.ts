import { IContext } from '@ecoflow/homebridge-ui/public/interfaces/contracts';
import {
  FormOnChangeCallback,
  IForm,
  IHomebridge,
  PluginConfig,
  PluginConfigSchema,
  PluginConfigSchemaDevices,
  PluginConfigSchemaEnum,
  PluginConfigSchemaObject,
  PluginDeviceConfig,
} from '@ecoflow/homebridge-ui/public/interfaces/homebridge';
import { ComponentRenderer } from '@ecoflow/homebridge-ui/public/renderers/componentRenderer';
import { PluginConfigDevicesRenderer } from '@ecoflow/homebridge-ui/public/renderers/pluginConfigDevicesRenderer';
import $ from 'jquery';

describe('PluginConfigDevicesRenderer', () => {
  let context: IContext;
  let homebridgeProviderMock: jest.Mocked<IHomebridge>;
  let formMock: jest.Mocked<IForm>;
  let configuration: PluginConfig;
  let configSchema: PluginConfigSchema;
  let devicesContainer: HTMLElement;
  let devicesTabsElement: HTMLElement;
  let devicesTabPanelsElement: HTMLElement;
  let renderer: PluginConfigDevicesRenderer;

  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).$ = $;
  });

  afterAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (global as any).$ = undefined;
  });

  beforeEach(() => {
    homebridgeProviderMock = {
      createForm: jest.fn(),
      updatePluginConfig: jest.fn(),
    } as unknown as jest.Mocked<IHomebridge>;
    formMock = { onChange: jest.fn(), end: jest.fn() } as unknown as jest.Mocked<IForm>;
    homebridgeProviderMock.createForm.mockReturnValue(formMock);
    configuration = {
      name: 'name1',
      platform: 'EcoFlowPluginPlatform',
      devices: [
        {
          name: 'First Device',
          model: 'Delta 2 Max',
          prop1: 'true',
          battery: {
            additionalCharacteristics: ['Battery Level, %'],
          } as unknown as string,
        },
        { name: 'Second Device', model: 'PowerStream', prop1: 'false' },
      ],
    };
    const modelSchema: PluginConfigSchemaEnum = {
      title: 'Model',
      default: 'Delta 2',
      required: true,
      enum: ['Delta 2', 'Delta 2 Max', 'PowerStream'],
    };
    const batterySchema = {
      properties: {
        additionalCharacteristics: {
          items: {
            title: 'Characteristic',
            type: 'string',
            enum: ['Battery Level, %', 'Input Consumption, W', 'Output Consumption, W'],
          },
        },
      },
    } as unknown as PluginConfigSchemaObject;
    const powerStreamSchema = {
      properties: {
        type: {
          title: 'Type',
          required: true,
          enum: ['600', '800'],
        },
        pvAdditionalCharacteristics: {
          title: 'PV Additional Characteristics',
          description: 'List of additional characteristics for PVs connected for PowerStream',
          items: {
            title: 'Characteristic',
            enum: ['Output Consumption, W'],
          },
        },
      },
    } as unknown as PluginConfigSchemaObject;
    const outletPlugSchema = {
      properties: {
        additionalCharacteristics: {
          items: {
            title: 'Characteristic',
            type: 'string',
            enum: ['Output Voltage, V', 'Output Current, A', 'Output Consumption, W'],
          },
        },
      },
    } as unknown as PluginConfigSchemaObject;
    configSchema = {
      schema: {
        properties: {
          name: { title: 'Name Title' },
          devices: {
            title: 'Devices',
            items: {
              properties: {
                name: {
                  title: 'Name',
                  required: true,
                  description: 'Name of the EcoFlow device',
                },
                model: modelSchema,
                prop1: {
                  title: 'Simulate',
                  default: 'true',
                  required: false,
                },
                battery: batterySchema,
                powerStream: powerStreamSchema,
                outlet: outletPlugSchema,
              },
            },
          },
        },
      },
    };
    context = { homebridgeProvider: homebridgeProviderMock, configuration, configSchema };

    devicesContainer = document.createElement('div');
    document.body.replaceChildren(devicesContainer);
    devicesTabsElement = document.createElement('ul');
    devicesTabsElement.id = 'devicesTabs';
    devicesTabPanelsElement = document.createElement('div');
    devicesTabPanelsElement.id = 'devicesPanels';
    devicesContainer.appendChild(devicesTabsElement);
    devicesContainer.appendChild(devicesTabPanelsElement);

    renderer = new PluginConfigDevicesRenderer(new ComponentRenderer());
  });

  function clone<TObject>(obj: TObject): TObject {
    return JSON.parse(JSON.stringify(obj));
  }

  function cleanupAdditionalProperties(expectedSchema: PluginConfigSchemaDevices, ...names: string[]): void {
    names.forEach(name => {
      delete expectedSchema.items.properties[name];
    });
  }

  describe('hideDeviceSettingsPerModel', () => {
    it('should contain proper hideDeviceSettingsPerModel configuration', () => {
      const expected = {
        'Delta 2': ['powerStream', 'outlet'],
        'Delta 2 Max': ['powerStream', 'outlet'],
        'Delta Pro': ['powerStream', 'outlet'],
        'Delta Pro 3': ['powerStream', 'outlet'],
        'Delta Pro Ultra': ['powerStream', 'outlet'],
        PowerStream: ['battery', 'outlet'],
        'Smart Plug': ['battery', 'powerStream'],
      };

      const actual = renderer.hideDeviceSettingsPerModel;

      expect(actual).toEqual(expected);
    });
  });

  describe('render', () => {
    function activateTab(index: number): void {
      const tab = $(devicesTabsElement).find(`#deviceTab${index}`);
      tab.trigger('click');
    }

    function removeDevice(index: number): void {
      const removeDevice = $(devicesTabPanelsElement).find(`#deviceTabPanelClose${index}`);
      removeDevice.trigger('click');
    }

    describe('tabsRendering', () => {
      it('should render tabs when there are 2 devices in configuration', () => {
        renderer.render(context);

        expect(devicesTabsElement).toMatchSnapshot();
      });

      it('should render tabs when there are no devices in configuration', () => {
        context.configuration.devices = [];
        renderer.render(context);

        expect(devicesTabsElement).toMatchSnapshot();
      });
    });

    describe('tabsNavigation', () => {
      let expectedSchema: PluginConfigSchemaDevices;

      beforeEach(() => {
        expectedSchema = clone(configSchema.schema.properties.devices);
        cleanupAdditionalProperties(expectedSchema, 'model');
      });

      it('should render form for first device in configuration by default', () => {
        cleanupAdditionalProperties(expectedSchema, 'powerStream', 'outlet');
        const expectedConfiguration = clone(configuration).devices[0];

        renderer.render(context);

        expect(homebridgeProviderMock.createForm).toHaveBeenCalledTimes(1);
        expect(homebridgeProviderMock.createForm).toHaveBeenCalledWith(
          { schema: expectedSchema.items },
          expectedConfiguration
        );
      });

      it('should render form for second device when second tab is clicked', () => {
        cleanupAdditionalProperties(expectedSchema, 'battery', 'outlet');
        const expectedConfiguration = clone(configuration).devices[1];

        renderer.render(context);
        homebridgeProviderMock.createForm.mockClear();
        activateTab(1);

        expect(homebridgeProviderMock.createForm).toHaveBeenCalledTimes(1);
        expect(homebridgeProviderMock.createForm).toHaveBeenCalledWith(
          { schema: expectedSchema.items },
          expectedConfiguration
        );
        expect(formMock.end).toHaveBeenCalledTimes(1);
      });
    });

    describe('tabPanelsRendering', () => {
      it('should render tab panels and activate first one when there are 2 devices in configuration', () => {
        renderer.render(context);

        expect(devicesTabPanelsElement).toMatchSnapshot();
      });

      it('should render tab panels when there are no devices in configuration', () => {
        context.configuration.devices = [];
        renderer.render(context);

        expect(devicesTabPanelsElement).toMatchSnapshot();
      });
    });

    describe('addNewDevice', () => {
      let expectedSchema: PluginConfigSchemaDevices;

      beforeEach(() => {
        expectedSchema = clone(configSchema.schema.properties.devices);
        cleanupAdditionalProperties(expectedSchema, 'model', 'powerStream', 'outlet');
        renderer.render(context);
        homebridgeProviderMock.createForm.mockClear();
      });

      it('should render form for new device when addDeviceTab is clicked', () => {
        const deviceTabAdd = $(devicesTabsElement).find('#deviceTabAdd');
        deviceTabAdd.trigger('click');

        expect(homebridgeProviderMock.createForm).toHaveBeenCalledTimes(1);
        expect(homebridgeProviderMock.createForm).toHaveBeenCalledWith(
          { schema: expectedSchema.items },
          { model: 'Delta 2', prop1: 'true' }
        );
        expect(formMock.end).toHaveBeenCalledTimes(2);
      });

      it('should render tab with default values when addDeviceTab is clicked', () => {
        const deviceTabAdd = $(devicesTabsElement).find('#deviceTabAdd');
        deviceTabAdd.trigger('click');

        expect(devicesContainer).toMatchSnapshot();
      });
    });

    describe('removeDevice', () => {
      beforeEach(() => {
        renderer.render(context);
      });

      it('should remove active tab when remove button is clicked', () => {
        activateTab(1);
        removeDevice(1);

        expect(devicesContainer).toMatchSnapshot();
      });

      it('should leave addNewDevice tab only when remove button is clicked for last device', () => {
        activateTab(1);
        removeDevice(1);
        removeDevice(0);

        expect(devicesContainer).toMatchSnapshot();
      });
    });

    describe('applyChanges', () => {
      let formOnChangeCallback: FormOnChangeCallback;

      beforeEach(() => {
        renderer.render(context);
        formOnChangeCallback = formMock.onChange.mock.calls[0][0];
      });

      it('should update changes when there are any made on form', async () => {
        await formOnChangeCallback({
          name: 'First Device Changed',
          prop1: 'false',
          battery: {
            additionalCharacteristics: ['Battery Level, %', 'Input Consumption, W'],
          } as unknown as string,
        } as unknown as PluginDeviceConfig);

        expect(homebridgeProviderMock.updatePluginConfig).toHaveBeenCalledTimes(1);
        expect(homebridgeProviderMock.updatePluginConfig).toHaveBeenCalledWith([
          {
            name: 'name1',
            devices: [
              {
                name: 'First Device Changed',
                model: 'Delta 2 Max',
                prop1: 'false',
                battery: {
                  additionalCharacteristics: ['Battery Level, %', 'Input Consumption, W'],
                } as unknown as string,
              },
              { name: 'Second Device', model: 'PowerStream', prop1: 'false' },
            ],
          },
        ]);
      });

      it('should update tab name when name was changed on form', async () => {
        await formOnChangeCallback({ name: 'First Device Changed' } as unknown as PluginDeviceConfig);
        const actual = $(devicesTabsElement).find('#deviceTab0').text();

        expect(actual).toBe('First Device Changed');
      });
    });

    describe('modelChanged', () => {
      let modelElement: JQuery;
      let expectedSchema: PluginConfigSchemaDevices;

      beforeEach(() => {});

      beforeEach(() => {
        expectedSchema = clone(configSchema.schema.properties.devices);
        cleanupAdditionalProperties(expectedSchema, 'model');

        renderer.render(context);
        homebridgeProviderMock.createForm.mockClear();
        modelElement = $(devicesTabPanelsElement).find('#devicemodel0');
      });

      it('should update tab name when name was changed on form', () => {
        modelElement.val('PowerStream');
        modelElement.trigger('change');

        cleanupAdditionalProperties(expectedSchema, 'battery', 'outlet');
        const expectedConfiguration = clone(configuration).devices[0];

        expect(homebridgeProviderMock.createForm).toHaveBeenCalledTimes(1);
        expect(homebridgeProviderMock.createForm).toHaveBeenCalledWith(
          { schema: expectedSchema.items },
          expectedConfiguration
        );
        expect(formMock.end).toHaveBeenCalledTimes(1);
      });
    });
  });
});
