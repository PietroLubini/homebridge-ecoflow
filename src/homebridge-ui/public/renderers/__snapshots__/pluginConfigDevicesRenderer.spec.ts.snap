// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginConfigDevicesRenderer render addNewDevice should render tab with default values when addDeviceTab is clicked 1`] = `
<div>
  <ul
    id="devicesTabs"
  >
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link "
        data-toggle="tab"
        href="#deviceTabPanel0"
        id="deviceTab0"
      >
        First Device
      </a>
      
      
    </li>
    
    
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link "
        data-toggle="tab"
        href="#deviceTabPanel1"
        id="deviceTab1"
      >
        Second Device
      </a>
      
      
    </li>
    
    
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link active"
        data-toggle="tab"
        href="#deviceTabPanel2"
        id="deviceTab2"
      >
        Device3
      </a>
      
      
    </li>
    
    
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link "
        data-toggle="tab"
        href="#deviceTabPanelAdd"
        id="deviceTabAdd"
      >
        Add Device
      </a>
      
      
    </li>
    
    
  </ul>
  <div
    id="devicesPanels"
  >
    
    
    <div
      class="tab-pane container fade  card card-body list-group-item"
      id="deviceTabPanel0"
    >
      
      
      <button
        class="close pull-right"
        id="deviceTabPanelClose0"
        type="button"
      >
        
        
        <span>
          ×
        </span>
        
        
        <span
          class="sr-only"
        >
          Close
        </span>
        
      
      </button>
      
    
      
    
      <div
        class="form-group"
      >
        
      
        <label
          class="required-label"
          for="devicemodel0"
        >
          Model
        </label>
        
      
        <select
          class="form-control"
          id="devicemodel0"
          required="true"
        >
          <option
            value="Delta 2"
          >
            Delta 2
          </option>
          <option
            value="Delta 2 Max"
          >
            Delta 2 Max
          </option>
          <option
            value="PowerStream"
          >
            PowerStream
          </option>
        </select>
        
    
      </div>
    </div>
    
  
    
    
    <div
      class="tab-pane container fade  card card-body list-group-item"
      id="deviceTabPanel1"
    >
      
      
      <button
        class="close pull-right"
        id="deviceTabPanelClose1"
        type="button"
      >
        
        
        <span>
          ×
        </span>
        
        
        <span
          class="sr-only"
        >
          Close
        </span>
        
      
      </button>
      
    
      
    
      <div
        class="form-group"
      >
        
      
        <label
          class="required-label"
          for="devicemodel1"
        >
          Model
        </label>
        
      
        <select
          class="form-control"
          id="devicemodel1"
          required="true"
        >
          <option
            value="Delta 2"
          >
            Delta 2
          </option>
          <option
            value="Delta 2 Max"
          >
            Delta 2 Max
          </option>
          <option
            value="PowerStream"
          >
            PowerStream
          </option>
        </select>
        
    
      </div>
    </div>
    
  
    
    
    <div
      class="tab-pane container fade in show active card card-body list-group-item"
      id="deviceTabPanel2"
    >
      
      
      <button
        class="close pull-right"
        id="deviceTabPanelClose2"
        type="button"
      >
        
        
        <span>
          ×
        </span>
        
        
        <span
          class="sr-only"
        >
          Close
        </span>
        
      
      </button>
      
    
      
    
      <div
        class="form-group"
      >
        
      
        <label
          class="required-label"
          for="devicemodel2"
        >
          Model
        </label>
        
      
        <select
          class="form-control"
          id="devicemodel2"
          required="true"
        >
          <option
            value="Delta 2"
          >
            Delta 2
          </option>
          <option
            value="Delta 2 Max"
          >
            Delta 2 Max
          </option>
          <option
            value="PowerStream"
          >
            PowerStream
          </option>
        </select>
        
    
      </div>
    </div>
    
  
  </div>
</div>
`;

exports[`PluginConfigDevicesRenderer render removeDevice should leave addNewDevice tab only when remove button is clicked for last device 1`] = `
<div>
  <ul
    id="devicesTabs"
  >
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link "
        data-toggle="tab"
        href="#deviceTabPanelAdd"
        id="deviceTabAdd"
      >
        Add Device
      </a>
      
      
    </li>
    
    
  </ul>
  <div
    id="devicesPanels"
  />
</div>
`;

exports[`PluginConfigDevicesRenderer render removeDevice should remove active tab when remove button is clicked 1`] = `
<div>
  <ul
    id="devicesTabs"
  >
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link active"
        data-toggle="tab"
        href="#deviceTabPanel0"
        id="deviceTab0"
      >
        First Device
      </a>
      
      
    </li>
    
    
    
      
    <li
      class="nav-item"
    >
      
        
      <a
        class="nav-link "
        data-toggle="tab"
        href="#deviceTabPanelAdd"
        id="deviceTabAdd"
      >
        Add Device
      </a>
      
      
    </li>
    
    
  </ul>
  <div
    id="devicesPanels"
  >
    
    
    <div
      class="tab-pane container fade in show active card card-body list-group-item"
      id="deviceTabPanel0"
    >
      
      
      <button
        class="close pull-right"
        id="deviceTabPanelClose0"
        type="button"
      >
        
        
        <span>
          ×
        </span>
        
        
        <span
          class="sr-only"
        >
          Close
        </span>
        
      
      </button>
      
    
      
    
      <div
        class="form-group"
      >
        
      
        <label
          class="required-label"
          for="devicemodel0"
        >
          Model
        </label>
        
      
        <select
          class="form-control"
          id="devicemodel0"
          required="true"
        >
          <option
            value="Delta 2"
          >
            Delta 2
          </option>
          <option
            value="Delta 2 Max"
          >
            Delta 2 Max
          </option>
          <option
            value="PowerStream"
          >
            PowerStream
          </option>
        </select>
        
    
      </div>
    </div>
    
  
  </div>
</div>
`;

exports[`PluginConfigDevicesRenderer render tabPanelsRendering should render tab panels and activate first one when there are 2 devices in configuration 1`] = `
<div
  id="devicesPanels"
>
  
    
  <div
    class="tab-pane container fade in show active card card-body list-group-item"
    id="deviceTabPanel0"
  >
    
      
    <button
      class="close pull-right"
      id="deviceTabPanelClose0"
      type="button"
    >
      
        
      <span>
        ×
      </span>
      
        
      <span
        class="sr-only"
      >
        Close
      </span>
      
      
    </button>
    
    
    
    
    <div
      class="form-group"
    >
      
      
      <label
        class="required-label"
        for="devicemodel0"
      >
        Model
      </label>
      
      
      <select
        class="form-control"
        id="devicemodel0"
        required="true"
      >
        <option
          value="Delta 2"
        >
          Delta 2
        </option>
        <option
          value="Delta 2 Max"
        >
          Delta 2 Max
        </option>
        <option
          value="PowerStream"
        >
          PowerStream
        </option>
      </select>
      
    
    </div>
  </div>
  
  
  
    
  <div
    class="tab-pane container fade  card card-body list-group-item"
    id="deviceTabPanel1"
  >
    
      
    <button
      class="close pull-right"
      id="deviceTabPanelClose1"
      type="button"
    >
      
        
      <span>
        ×
      </span>
      
        
      <span
        class="sr-only"
      >
        Close
      </span>
      
      
    </button>
    
    
    
    
    <div
      class="form-group"
    >
      
      
      <label
        class="required-label"
        for="devicemodel1"
      >
        Model
      </label>
      
      
      <select
        class="form-control"
        id="devicemodel1"
        required="true"
      >
        <option
          value="Delta 2"
        >
          Delta 2
        </option>
        <option
          value="Delta 2 Max"
        >
          Delta 2 Max
        </option>
        <option
          value="PowerStream"
        >
          PowerStream
        </option>
      </select>
      
    
    </div>
  </div>
  
  
</div>
`;

exports[`PluginConfigDevicesRenderer render tabPanelsRendering should render tab panels when there are no devices in configuration 1`] = `
<div
  id="devicesPanels"
/>
`;

exports[`PluginConfigDevicesRenderer render tabsRendering should render tabs when there are 2 devices in configuration 1`] = `
<ul
  id="devicesTabs"
>
  
      
  <li
    class="nav-item"
  >
    
        
    <a
      class="nav-link active"
      data-toggle="tab"
      href="#deviceTabPanel0"
      id="deviceTab0"
    >
      First Device
    </a>
    
      
  </li>
  
    
  
      
  <li
    class="nav-item"
  >
    
        
    <a
      class="nav-link "
      data-toggle="tab"
      href="#deviceTabPanel1"
      id="deviceTab1"
    >
      Second Device
    </a>
    
      
  </li>
  
    
  
      
  <li
    class="nav-item"
  >
    
        
    <a
      class="nav-link "
      data-toggle="tab"
      href="#deviceTabPanelAdd"
      id="deviceTabAdd"
    >
      Add Device
    </a>
    
      
  </li>
  
    
</ul>
`;

exports[`PluginConfigDevicesRenderer render tabsRendering should render tabs when there are no devices in configuration 1`] = `
<ul
  id="devicesTabs"
>
  
      
  <li
    class="nav-item"
  >
    
        
    <a
      class="nav-link "
      data-toggle="tab"
      href="#deviceTabPanelAdd"
      id="deviceTabAdd"
    >
      Add Device
    </a>
    
      
  </li>
  
    
</ul>
`;
